cpp/util/etcd.cc

```diff
@@ -803,101 +737,81 @@ EtcdClient::~EtcdClient() {
 }
 
 
-bool EtcdClient::MaybeUpdateLeader(const libevent::HttpRequest& req,
-                                   Request* etcd_req) {
-  // We're talking to the leader, get back to normal processing...
-  if (req.GetResponseCode() != 307) {
-    return false;
-  }
-
-  const char* const location(
-      CHECK_NOTNULL(evhttp_find_header(req.GetInputHeaders(), "location")));
+void EtcdClient::FetchDone(Request* etcd_req, util::Task* task) {
+  VLOG(2) << "EtcdClient::FetchDone: " << task->status();
 
-  const unique_ptr<evhttp_uri, void (*)(evhttp_uri*)> uri(
-      evhttp_uri_parse(location), &evhttp_uri_free);
-  CHECK(uri);
-  LOG(INFO) << "etcd leader: " << evhttp_uri_get_host(uri.get()) << ":"
-            << evhttp_uri_get_port(uri.get());
-
-  // Update the last known leader, and retry the request on the new
-  // leader.
-  etcd_req->Run(UpdateLeader(evhttp_uri_get_host(uri.get()),
-                             evhttp_uri_get_port(uri.get())));
+  if (!task->status().ok()) {
+    // TODO(pphaneuf): If there is a connection problem, we should
+    // re-do the request on another node.
+    etcd_req->task_->Return(task->status());
+    return;
+  } else {
+    VLOG(2) << "response:\n" << etcd_req->resp_;
+  }
 
-  return true;
-}
+  if (etcd_req->resp_.status_code == 307) {
+    UrlFetcher::Headers::const_iterator it(
+        etcd_req->resp_.headers.find("location"));
 
+    if (it == etcd_req->resp_.headers.end()) {
+      etcd_req->task_->Return(
+          Status(util::error::INTERNAL,
+                 "etcd returned a redirect without a Location header?"));
+      return;
+    }
 
-void EtcdClient::RequestDone(const shared_ptr<libevent::HttpRequest>& req,
-                             Request* etcd_req) {
-  // The HttpRequest object will be invalid as soon as we return, so
-  // forget about it now. It's too late to cancel, anyway.
-  etcd_req->Reset();
+    URL url(it->second);
+    if (url.Host().empty() || url.Port() == 0) {
+      etcd_req->task_->Return(
+          Status(util::error::INTERNAL,
+                 "could not parse Location header from etcd: " + it->second));
```

Comment:
We haven't been totally consistent, but as those strings aren't proper sentence (and can be put inside of other messages, etc), I prefer to keep them un-capitalised (sort of following internal Google style).